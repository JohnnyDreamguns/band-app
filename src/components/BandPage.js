import React from 'react';
import { Link } from 'react-router-dom';
const BandPage = ({ band, albums, loading }) => (
  <div>
    {loading && <div className="loader"></div>}

    {band && (
      <div>
        <h2>{band.name}</h2>
        <p>{band.bio}</p>
        <p>Formed: {band.formed}</p>
        <br />

        <h3>Albums</h3>
        {albums &&
          albums.map(album => (
            <p key={album.id}>
              <Link to={`/album/${album.id}`}>{album.name}</Link>
            </p>
          ))}

        <br />
        <p>
          Back to <Link to="/">Home</Link>
        </p>
      </div>
    )}
  </div>
);

export default React.memo(BandPage);
