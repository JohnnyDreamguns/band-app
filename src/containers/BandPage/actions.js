import {
  SET_BAND_ID,
  SET_LOADING,
  CLEAR_BAND_PAGE,
  SET_ALBUM_IDS
} from './constants';
import { doRequestMulti } from '../../services/api';
import { setBands, setAlbums, setTotalNumAlbums } from '../../actions/data';
import { setError } from '../../actions/messages';
import { catchErrors } from '../../utils/catchErrors';

export const getBandById = payload => {
  return async dispatch => {
    dispatch(setLoading(true));

    const { data, error } = await catchErrors(
      doRequestMulti(
        { endpoint: 'getBandById', params: { id: payload } },
        { endpoint: 'getAlbumsByBandId', params: { id: payload } }
      )
    );

    if (error) {
      dispatch(setError('There was a problem loading the data'));
    } else {
      dispatch(setBands(data[0].results));
      dispatch(setBandId(parseInt(Object.keys(data[0].results)[0])));

      dispatch(setAlbums(data[1].results));
      dispatch(setTotalNumAlbums(data[1].totalResults));
      dispatch(setAlbumIds(Object.keys(data[1].results).map(x => +x)));
    }

    dispatch(setLoading(false));
  };
};

export const setBandId = payload => ({
  type: SET_BAND_ID,
  payload
});

export const setAlbumIds = payload => ({
  type: SET_ALBUM_IDS,
  payload
});

export const setLoading = payload => ({
  type: SET_LOADING,
  payload
});

export const clearBandpage = () => ({
  type: CLEAR_BAND_PAGE
});
