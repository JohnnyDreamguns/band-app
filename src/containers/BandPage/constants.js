export const GET_BAND_BY_ID = 'bands-app/BandPage/GET_BAND_BY_ID';
export const SET_BAND_ID = 'bands-app/BandPage/SET_BAND_ID';
export const CLEAR_BAND_PAGE = 'bands-app/BandPage/CLEAR_BAND';
export const SET_ALBUM_IDS = 'bands-app/BandPage/SET_ALBUM_IDS';
export const SET_LOADING = 'bands-app/BandPage/SET_LOADING';
