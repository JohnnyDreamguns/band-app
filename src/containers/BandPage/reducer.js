import produce from 'immer';
import {
  SET_BAND_ID,
  CLEAR_BAND_PAGE,
  SET_LOADING,
  SET_ALBUM_IDS
} from './constants';

// The initial state of the App
export const initialState = {
  bandId: undefined,
  loading: false,
  albums: []
};

/* eslint-disable default-case, no-param-reassign */
const bandReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case SET_BAND_ID:
        draft.bandId = action.payload;
        break;

      case CLEAR_BAND_PAGE:
        draft.bandId = undefined;
        break;

      case SET_LOADING:
        draft.loading = action.payload;
        break;

      case SET_ALBUM_IDS:
        draft.albums = action.payload;
        break;
    }
  });

export default bandReducer;
