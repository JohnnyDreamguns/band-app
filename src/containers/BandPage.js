import React from 'react';
import BandPage from '../components/BandPage';
import { useBandData } from '../hooks/useBandData';

const BandPageContainer = props => {
  const { band, loading, albums, useFetchBandPageData } = useBandData();
  const { bandId } = props.match.params;

  useFetchBandPageData(bandId);

  const componentProps = {
    band,
    albums,
    loading
  };

  return <BandPage {...componentProps} />;
};

export default BandPageContainer;
