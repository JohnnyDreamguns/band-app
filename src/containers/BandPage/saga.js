import { call, put, takeLatest, all } from 'redux-saga/effects';
import { GET_BAND_BY_ID } from './constants';
import { setLoading, setBandId, setAlbumIds } from './actions';
import { setError } from '../../actions/messages';
import { setBands, setAlbums, setTotalNumAlbums } from '../../actions/data';

import { request } from '../../utils/request';

export function* getBandByIdSaga(action) {
  const requestURL = `http://localhost:3001/bands`;

  try {
    yield put(setLoading(true));

    const bandId = action.payload;

    const [band, albums] = yield all([
      call(request, `${requestURL}/${bandId}`),
      call(request, `${requestURL}/${bandId}/albums`)
    ]);

    yield put(setBands(band.results));
    yield put(setBandId(parseInt(Object.keys(band.results)[0])));

    yield put(setAlbums(albums.results));
    yield put(setTotalNumAlbums(albums.totalResults));
    yield put(setAlbumIds(Object.keys(albums.results).map(x => +x)));

    yield put(setLoading(false));
  } catch (err) {
    yield put(setLoading(false));
    yield put(setError('There was a problem loading the band data'));
  }
}

export default function* bandpageDataSaga() {
  yield takeLatest(GET_BAND_BY_ID, getBandByIdSaga);
}
