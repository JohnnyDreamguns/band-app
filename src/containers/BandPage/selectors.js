import { initialState } from './reducer';
import { selectBandsData, selectAlbumsData } from '../../selectors/data';

export const bandRoot = state => state.bandPage || initialState;

export const selectBandId = state => bandRoot(state)['bandId'];

export const selectBand = state => {
  const bandId = selectBandId(state);
  if (!bandId) return undefined;
  return selectBandsData(state)[`${bandId}`];
};

export const selectLoading = state => bandRoot(state)['loading'];

export const selectAlbumIds = state => bandRoot(state)['albums'];

export const selectAlbums = state => {
  const albumIds = selectAlbumIds(state);
  if (!albumIds) return [];

  const albums = albumIds
    .filter(x => selectAlbumsData(state)[x])
    .map(albumId => selectAlbumsData(state)[`${albumId}`]);

  return albums;
};
