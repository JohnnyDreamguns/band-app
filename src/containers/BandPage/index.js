import React, { useEffect } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { getBandById, clearBandpage } from './actions';
import { useInjectReducer } from '../../utils/injectReducer';
import { selectLoading, selectBand, selectAlbums } from './selectors';
import reducer from './reducer';

const key = 'bandPage';

const BandPage = ({
  band,
  albums,
  bandId,
  getBandById,
  clearBandpage,
  loading
}) => {
  useInjectReducer({ key, reducer });

  useEffect(() => {
    getBandById(bandId);

    return () => {
      clearBandpage();
    };
  }, [getBandById, clearBandpage, bandId]);

  return (
    <div>
      {loading && <div className="loader"></div>}

      {band && (
        <div>
          <h2>{band.name}</h2>
          <p>{band.bio}</p>
          <p>Formed: {band.formed}</p>
          <br />

          <h3>Albums</h3>
          {albums &&
            albums.map(album => (
              <p key={album.id}>
                <Link to={`/album/${album.id}`}>{album.name}</Link>
              </p>
            ))}

          <br />
          <p>
            Back to <Link to="/">Home</Link>
          </p>
        </div>
      )}
    </div>
  );
};

export const mapStateToProps = (state, ownProps) => {
  const { bandId } = ownProps.match.params;

  return {
    bandId,
    loading: selectLoading(state),
    band: selectBand(state),
    albums: selectAlbums(state)
  };
};

export const mapDispatchToProps = dispatch => ({
  getBandById: bandId => dispatch(getBandById(bandId)),
  clearBandpage: () => dispatch(clearBandpage())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(BandPage);
