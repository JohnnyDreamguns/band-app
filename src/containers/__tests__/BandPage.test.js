import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import TestRenderer from 'react-test-renderer';
import BandPageContainer from '../BandPage';
import BandPage from '../../components/BandPage';

const mockUseFetchBandPageData = jest.fn();

jest.mock('../../hooks/useBandData', () => ({
  useBandData: () => ({
    band: 'mock band',
    albumIds: 'mock albumIds',
    albums: [{ id: 1, name: 'mock album' }],
    loading: true,
    setLoading: 'mock setLoading',
    setBandId: 'mock setBandId',
    setAlbumKeys: 'mock setAlbumKeys',
    clearBandId: 'mock clearBandId',
    fetchBandPageData: 'mock fetchBandPageData',
    useFetchBandPageData: mockUseFetchBandPageData
  })
}));

jest.mock('../../hooks/useGlobalData', () => ({
  useGlobalData: () => ({
    setGlobalAlbums: () => {},
    setGlobalBands: () => {}
  })
}));

describe('BandPage container', () => {
  const props = {
    match: { params: { bandId: 2 } }
  };

  const testRenderer = TestRenderer.create(
    <BrowserRouter>
      <BandPageContainer {...props} />
    </BrowserRouter>
  );
  const testInstance = testRenderer.root;

  it('should match the props sent to the component', () => {
    expect(testInstance.findByType(BandPage.type).props).toMatchSnapshot();
  });

  it('should call mockUseFetchAlbumPageData', () => {
    expect(mockUseFetchBandPageData).toHaveBeenCalledTimes(1);
  });
});
