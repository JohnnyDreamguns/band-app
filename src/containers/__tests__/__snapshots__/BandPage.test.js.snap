// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BandPage container should match the props sent to the component 1`] = `
Object {
  "albums": Array [
    Object {
      "id": 1,
      "name": "mock album",
    },
  ],
  "band": "mock band",
  "loading": true,
}
`;
